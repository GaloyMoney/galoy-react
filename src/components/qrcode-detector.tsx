import { Html5Qrcode, Html5QrcodeSupportedFormats } from "html5-qrcode"
import { memo, useEffect, useRef, useState } from "react"

import { Spinner } from "./spinner"

type Props = {
  autoStart?: boolean
  startText: string
  stopText: string
  onCodeDetected: (barcodeValue: string) => any
  onValidCode: (data: any) => Promise<void>
}

const QRCodeDetecorComponent = ({
  autoStart = false,
  startText,
  stopText,
  onCodeDetected,
  onValidCode,
}: Props) => {
  const [detecting, setDetecting] = useState<boolean>(autoStart)
  const [cameraReady, setCameraReady] = useState<boolean>(false)
  const qrCodeRef = useRef<Html5Qrcode | null>(null)

  useEffect(() => {
    if (detecting) {
      qrCodeRef.current =
        qrCodeRef.current ||
        new Html5Qrcode("qrCodeCamera", {
          formatsToSupport: [Html5QrcodeSupportedFormats.QR_CODE],
          verbose: false,
        })

      const detectBarcode = async () => {
        const onScanSuccess = async (decodedText: string) => {
          const parsedResult = onCodeDetected(decodedText)

          if (parsedResult) {
            onValidCode(parsedResult)
            await qrCodeRef.current?.stop()
            setDetecting(false)
          }
        }

        await qrCodeRef.current?.start(
          { facingMode: "environment" },
          { fps: 10, qrbox: 400 },
          onScanSuccess,
          () => {
            // Do nothing for invalid scans
          },
        )
        setCameraReady(true)
      }
      detectBarcode()
    }
  }, [detecting, onCodeDetected, onValidCode])

  const startDetecting = () => {
    setDetecting(true)
  }

  const stopDetecting = async () => {
    await qrCodeRef.current?.stop()
    setDetecting(false)
  }

  return (
    <div className="qrcode-detector">
      {detecting ? (
        <div className="qrcode-camera">
          <div id="qrCodeCamera"></div>
          {cameraReady ? (
            <div className="scan-stop link" onClick={stopDetecting}>
              {stopText}
            </div>
          ) : (
            <Spinner size="big" />
          )}
        </div>
      ) : (
        <div className="scan-start link" onClick={startDetecting}>
          <i aria-hidden className="fas fa-qrcode"></i>
          {startText}
        </div>
      )}
    </div>
  )
}

export const QRCodeDetecor = memo(QRCodeDetecorComponent)
