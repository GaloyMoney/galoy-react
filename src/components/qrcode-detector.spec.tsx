import { render } from "@testing-library/react"
import { QRCodeDetecor } from "./qrcode-detector"

describe("SatSymbol", () => {
  it("renders and matches snapshot", () => {
    const { asFragment } = render(
      <QRCodeDetecor
        startText="start"
        stopText="stop"
        onCodeDetected={jest.fn()}
        onValidCode={jest.fn()}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
