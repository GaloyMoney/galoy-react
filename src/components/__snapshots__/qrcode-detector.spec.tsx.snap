// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SatSymbol renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="qrcode-detector"
  >
    <div
      class="scan-start link"
    >
      <i
        aria-hidden="true"
        class="fas fa-qrcode"
      />
      start
    </div>
  </div>
</DocumentFragment>
`;
